File: cukcuk clone/app/src/main/java/vn/com/misa/cukcukstarterclone/data/source/local/database/AppDatabase.java
package vn.com.misa.cukcukstarterclone.data.source.local.database;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import static vn.com.misa.cukcukstarterclone.utils.Queries.CREATE_CART_ITEM_TABLE;
import static vn.com.misa.cukcukstarterclone.utils.Queries.CREATE_CART_TABLE;
import static vn.com.misa.cukcukstarterclone.utils.Queries.CREATE_MENU_GROUP_TABLE;
import static vn.com.misa.cukcukstarterclone.utils.Queries.CREATE_MENU_TABLE;
import static vn.com.misa.cukcukstarterclone.utils.Queries.CREATE_ORDER_TABLE;
import static vn.com.misa.cukcukstarterclone.utils.Queries.DATABASE_NAME;
import static vn.com.misa.cukcukstarterclone.utils.Queries.DATABASE_VERSION;
import static vn.com.misa.cukcukstarterclone.utils.Queries.DROP_CART_ITEM_TABLE;
import static vn.com.misa.cukcukstarterclone.utils.Queries.DROP_CART_TABLE;
import static vn.com.misa.cukcukstarterclone.utils.Queries.DROP_MENU_GROUP_TABLE;
import static vn.com.misa.cukcukstarterclone.utils.Queries.DROP_MENU_TABLE;
import static vn.com.misa.cukcukstarterclone.utils.Queries.DROP_ORDER_TABLE;

/**
 * - Mục đích Class:
 * - Sử dụng khi:
 *
 * @created_by KhanhNQ on 07-Jan-21.
 */
public class AppDatabase extends SQLiteOpenHelper {

    public static volatile AppDatabase instance = null;

    public static AppDatabase getInstance(Context context) {
        if (null == instance) {
            instance = new AppDatabase(context, DATABASE_NAME, null, DATABASE_VERSION);
        }
        return instance;
    }

    private AppDatabase(Context context, String name, SQLiteDatabase.CursorFactory factory, int version) {
        super(context, name, factory, version);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL(CREATE_MENU_GROUP_TABLE);
        db.execSQL(CREATE_MENU_TABLE);
        db.execSQL(CREATE_CART_TABLE);
        db.execSQL(CREATE_CART_ITEM_TABLE);
        db.execSQL(CREATE_ORDER_TABLE);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        db.execSQL(DROP_MENU_GROUP_TABLE);
        db.execSQL(DROP_MENU_TABLE);
        db.execSQL(DROP_CART_TABLE);
        db.execSQL(DROP_CART_ITEM_TABLE);
        db.execSQL(DROP_ORDER_TABLE);

        onCreate(db);
    }
}

File: app/src/main/java/com/misa/fresher/data/source/local/database/AppDatabase.kt
package com.misa.fresher.data.source.local.database

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import com.misa.fresher.common.Constant
import com.misa.fresher.common.FakeData
import com.misa.fresher.data.entity.*

/**
 * Helper chứa hàm get database, ngoài ra còn sử dụng để tạo, cập nhật database
 *
 * @author Nguyễn Công Chính
 * @since 3/24/2022
 *
 * @version 2
 * @updated 3/24/2022: Tạo class
 * @updated 3/28/2022: Thêm dữ liệu giả ngay lúc khởi tạo csdl
 */
class AppDatabase private constructor(context: Context)
    : SQLiteOpenHelper(context, Constant.DATABASE_NAME, null, Constant.DATABASE_VERSION) {

    override fun onCreate(db: SQLiteDatabase?) {
        db?.run {
            execSQL(Constant.CREATE_CATEGORY_TABLE)
            execSQL(Constant.CREATE_CUSTOMER_TABLE)
            execSQL(Constant.CREATE_PRODUCT_COLOR_TABLE)
            execSQL(Constant.CREATE_PRODUCT_SIZE_TABLE)
            execSQL(Constant.CREATE_PRODUCT_UNIT_TABLE)
            execSQL(Constant.CREATE_PRODUCT_ITEM_TABLE)
            execSQL(Constant.CREATE_PRODUCT_TABLE)
            execSQL(Constant.CREATE_PRODUCT_ITEM_BILL_TABLE)
            execSQL(Constant.CREATE_BILL_TABLE)

            setDefaultCategory(this)
            setDefaultCustomer(this)
            setDefaultColor(this)
            setDefaultSize(this)
            setDefaultUnit(this)
            setDefaultProduct(this)
        }
    }

    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
        db?.run {
            execSQL(Constant.DROP_BILL_TABLE)
            execSQL(Constant.DROP_PRODUCT_ITEM_BILL_TABLE)
            execSQL(Constant.DROP_PRODUCT_TABLE)
            execSQL(Constant.DROP_PRODUCT_ITEM_TABLE)
            execSQL(Constant.DROP_PRODUCT_UNIT_TABLE)
            execSQL(Constant.DROP_PRODUCT_SIZE_TABLE)
            execSQL(Constant.DROP_PRODUCT_COLOR_TABLE)
            execSQL(Constant.DROP_CUSTOMER_TABLE)
            execSQL(Constant.DROP_CATEGORY_TABLE)
        }

        onCreate(db)
    }

    private fun setDefaultCategory(db: SQLiteDatabase) {
        db.beginTransaction()

        var isSuccess = true
        for (item in FakeData.category) {
            val values = item.getContentValues()
            if (db.insertWithOnConflict(
                    Category.TABLE_NAME,
                    null,
                    values,
                    SQLiteDatabase.CONFLICT_REPLACE) <= 0) {
                isSuccess = false
                break
            }
        }
        if (isSuccess) {
            db.setTransactionSuccessful()
        }
        db.endTransaction()
    }

    private fun setDefaultCustomer(db: SQLiteDatabase) {
        db.beginTransaction()

        var isSuccess = true
        for (item in FakeData.customers) {
            val values = item.getContentValues()
            if (db.insertWithOnConflict(
                    Customer.TABLE_NAME,
                    null,
                    values,
                    SQLiteDatabase.CONFLICT_REPLACE) <= 0) {
                isSuccess = false
                break
            }
        }
        if (isSuccess) {
            db.setTransactionSuccessful()
        }
        db.endTransaction()
    }

    private fun setDefaultColor(db: SQLiteDatabase) {
        db.beginTransaction()

        var isSuccess = true
        for (item in FakeData.colors) {
            val values = item.getContentValues()
            if (db.insertWithOnConflict(
                    ProductColor.TABLE_NAME,
                    null,
                    values,
                    SQLiteDatabase.CONFLICT_REPLACE) <= 0) {
                isSuccess = false
                break
            }
        }
        if (isSuccess) {
            db.setTransactionSuccessful()
        }
        db.endTransaction()
    }

    private fun setDefaultSize(db: SQLiteDatabase) {
        db.beginTransaction()

        var isSuccess = true
        for (item in FakeData.sizes) {
            val values = item.getContentValues()
            if (db.insertWithOnConflict(
                    ProductSize.TABLE_NAME,
                    null,
                    values,
                    SQLiteDatabase.CONFLICT_REPLACE) <= 0) {
                isSuccess = false
                break
            }
        }
        if (isSuccess) {
            db.setTransactionSuccessful()
        }
        db.endTransaction()
    }

    private fun setDefaultUnit(db: SQLiteDatabase) {
        db.beginTransaction()

        var isSuccess = true
        for (item in FakeData.units) {
            val values = item.getContentValues()
            if (db.insertWithOnConflict(
                    ProductUnit.TABLE_NAME,
                    null,
                    values,
                    SQLiteDatabase.CONFLICT_REPLACE) <= 0) {
                isSuccess = false
                break
            }
        }
        if (isSuccess) {
            db.setTransactionSuccessful()
        }
        db.endTransaction()
    }

    private fun setDefaultProduct(db: SQLiteDatabase) {
        db.beginTransaction()

        var isSuccess = true
        for (item in FakeData.products) {
            val values = item.getContentValues()
            if (db.insertWithOnConflict(
                    Product.TABLE_NAME,
                    null,
                    values,
                    SQLiteDatabase.CONFLICT_REPLACE) > 0) {
                for (ite in item.items) {
                    val itemValues = ite.getContentValues()
                    itemValues.put(ProductItem.PRODUCT_ID, item.id)
                    if (db.insertWithOnConflict(
                            ProductItem.TABLE_NAME,
                            null,
                            itemValues,
                            SQLiteDatabase.CONFLICT_REPLACE) <= 0) {
                        isSuccess = false
                        break
                    }
                }
            } else {
                isSuccess = false
                break

            }
        }
        if (isSuccess) {
            db.setTransactionSuccessful()
        }
        db.endTransaction()
    }

    companion object {
        private var appDatabase: AppDatabase? = null

        fun getInstance(context: Context): AppDatabase =
            appDatabase ?: AppDatabase(context).also { appDatabase = it }
    }
}